### PR TITLE
model modification

### PR DESCRIPTION
**Pull request by wmutian**
Dear author,

亲爱的作者，您好！

Thank you for your outstanding contribution. I have modified your work titled "glucose.ipynb," transforming it into "model_training_program.py," which I am preparing to upload soon. Additionally, I have saved the trained model as "ppg_glucose_model.h5," which you can utilize in "model_caller.py" for further testing and enhancements.

感谢您的杰出贡献。我已对您的作品“glucose.ipynb”进行了一些修改，并将其修改为“model_training_program.py”，准备不久后上传。此外，我已将训练好的模型保存为“ppg_glucose_model.h5”，您可以在“model_caller.py”中利用此模型进行进一步的测试和优化。

In terms of modifications, I have made several key changes: the input layer neurons have been increased from 16 to 64, the hidden layer neurons have been adjusted from 8 to 32, the batch size has been modified from 128 to 512, the iteration count has been increased from 30 to 500, and the epochs have been extended from 200 to 1000.

在修改时，我做了几个关键性调整：输入层神经元从16增加到64，隐藏层神经元从8调整为32，批处理大小从128修改为512，迭代次数从30增加到500，训练周期从200延长至1000。

The trained model, named "ppg_glucose_model.h5," can be invoked using "model_caller.py." However, please make sure to update the path to the model file before running the program. When running the program, please ensure that the input file is in .txt format and follows the specified sequence: "label-red PPG-infrared PPG-...-label-red PPG-...". Each interval between labels should contain 16 lines of data, with 8 lines dedicated to red PPG data and 8 lines to infrared PPG data. Upon inputting this data file, the program will provide predicted glucose levels (remember to specify the file save path in program).

这个名为“ppg_glucose_model.h5”的训练模型可通过“model_caller.py”进行调用。但请在运行程序前确保更新程序中模型文件的路径。运行程序时，请确保输入文件为.txt格式，并按照指定顺序：“label-red PPG-infrared PPG-...-label-red PPG-...”排列。每个标签间隔应包含16行数据，其中8行是红光光电容积脉搏波（PPG）数据，8行是红外光电容积脉搏波（PPG）数据。输入此数据文件后，程序将提供预测的血糖水平（请记得在程序中指定文件保存路径）。

The test results, presented in Figures 1 through 5, demonstrate a  correlation between the predicted glucose levels and the actual values, which were obtained using a glucometer (Sinocare, Anwen+ Pro). Based on these exciting results, I am submitting this pull request and inforimg this to you. Your valuable opinions are welcome.

图1至图5中展示了测试结果，这些结果证明了预测的血糖水平与实际值（使用"三诺安稳+"血糖仪测得）之间存在相关性。基于这些令人兴奋的结果，我现提交此请求并通知您。欢迎您的宝贵意见。

### CODE REVIEW

**Comment by vvvirs** [comment]
Hello, I have also been doing research on blood glucose estimation based on PPG recently, and I am very glad to see that your test results can match the results of the blood glucose meter. However, my current research has failed to achieve your effect. Could you please tell me the data set collection method and pre-processing method used in model training? I would appreciate your reply.

您好，我最近也在做基于PPG估计血糖的研究，很高兴能看到您的测试结果能匹配上血糖仪的结果。然而我目前研究未能达到您的效果，请问您方便透露一下模型训练使用的数据集采集方法以及预处理的方法吗？如能答复，我将不胜感激！

**Comment by wmutian** [comment]
Hello! Thank you very much for your attention and question. I'm very glad that you are interested in my contribution and understand the challenges you are currently facing in your research.
您好！非常感谢您的关注和提问。我很高兴您对我的贡献感兴趣，也理解您目前在研究中遇到的挑战。

The data I used for model training all came from the original author's dataset. When validating the model, I used data that collected by myself, which were obtained through the MAX30102 pulse oximeter module.
我训练时使用的数据全部来源于原作者提供的数据集。在验证模型时，我使用了自己采集的数据，这些数据是通过MAX30102脉搏血氧模块采集的。



The specific procedure is to place the pad of the index finger on the light source of the module. The infrared (IR) and red light (red) data are then read from the module's FIFO buffer every 20 ms. Additionally, a marker row is inserted every 16 rows in the data ("11551155"). The data is saved in a TXT format file. The file name can be modified according to the format of the original author's dataset if necessary. If it is only used for prediction, modification is not required. After the data collection is complete, I manually delete the data before the first marker row and the last marker row as well as the data after it, to ensure that the data conforms to the input format required by the model.
具体操作是将食指指腹贴在模块的光源上，从模块的FIFO缓冲区内读取红外（ir）和红光（red）数据，每20毫秒读取一次，每隔16行数据插入一行标志行（“11551155”），保存为txt文本格式。文件名可以按原作者的数据集格式进行修改，如果仅用来预测可以不修改。采集完成后，我会手动删除第一行标志行之前的数据以及最后一个标志行及其之后的数据，以确保数据符合模型输入的格式。

屏幕截图 2025-02-14 160847

After that, the TXT file containing the data can be opened in “model_caller.py” for blood glucose prediction. In “model_caller.py”, the collected data is first denoised using the “3σ rule” to reduce the impact of environmental interference and other factors on the data. Subsequently, the data undergoes z-score normalization to enhance the prediction accuracy of the model.
之后，这个包含数据的txt文件便可以在“model_caller.py”中打开进行血糖的预测。在“model_caller.py”中，对采集到的数据按照“3σ准则”进行了去噪处理，以减少环境干扰等因素对数据的影响；然后对数据进行了z-score归一化处理，以提高模型的预测精度。

The core part of the data acquisition program is as follows for your reference:
数据采集程序核心部分如下，供您参考：

```
extern uint16_t fifo_red;			//max30102测得的红光（波长660 nm）光强
extern uint16_t fifo_ir;			//max30102测得的红外光（波长880 nm）光强
int main(void)
{
	int c=0,k=11551155;
	char result[8];  

	SystemClock_Config(); 		      //初始化
  	HAL_Init();
  	MX_GPIO_Init();		
	MX_USART1_UART_Init();
	MX_I2C1_Init();
	MX_I2C2_Init();
	max30102_init();
	max30102_fir_init();

 	float max30102_data[2], fir_output[2];
  	printf("Max30102 Init\r\n");
	
	HAL_UART_Receive_IT(&huart1,&aRxBuffer1,1);	// 启用UART中断接收
	HAL_Delay(500);					// 延迟500毫秒  
	printf("%d\n",k);
	
  while (1)								       		
  {
		max30102_read_fifo();
		printf("%d\n",fifo_red);	//红光（red）光强
		printf("%d\n",fifo_ir);	        //红外（ir）光强
		HAL_Delay(20);
		c++;
		if(c%8==0)
		{
			printf("%d\n",k);      //每隔16行数据插入标志行
		}			
  }
}
```

I understand the situation you described. I have also tried multiple times to reproduce my own results. Even with the same program and the same dataset, the models obtained from each training show significant differences. I think this is due to the inherent randomness in the neural network training process, such as the random initialization of initial weights and the stochastic nature of gradient descent during training.
我理解您描述的情况，我也多次尝试复现自己的结果，即使是相同的程序和相同的数据集，每次训练得到的模型均有较大差异。我认为这种情况是神经网络训练过程的随机性造成的，比如初始权重的随机初始化、训练过程中的随机梯度下降等。

Therefore, I suggest that you directly use the “ppg_glucose_model.h5” model file that I have uploaded. This model is a relatively better-performing version after multiple rounds of training and validation, and it may provide some reference and assistance for your research. My contribution only demonstrates the feasibility of predicting blood glucose levels using a multilayer perceptron (MLP). If your research requires more precise prediction of blood glucose through PPG, I recommend exploring more advanced neural network structures and algorithms, such as the CNN-LSTM that my team is currently investigating.
因此，我建议您直接调用我上传的“ppg_glucose_model.h5”模型文件。该模型是我经过多次训练和验证后，预测效果相对较好的版本，或许可以为您的研究提供一定的参考和帮助。我的这一部分贡献只能说明使用多层感知机（MLP）预测血糖具有一定的可行性，如果您的研究需要更精确地通过PPG预测血糖，建议您采用更高级的神经网络架构和算法（比如我所在的团队正在探索CNN-LSTM）。

I hope the above content is helpful to you. If you have any other questions during your research, feel free to communicate with me at any time. Thank you again for your attention and support!
希望以上对内容您有所帮助。如果您在研究过程中还有其他问题，欢迎随时与我交流。再次感谢您的关注和支持！

**Comment by vvvirs** [comment]
Thank you for your reply! When I was training the model using the original author's data, I found that the data set was not the original PPG data, as shown in the figure.

感谢您的答复！当我在使用原作者的数据训练模型时，我发现数据集并非是原始的PPG数据，如下图所示。

image


And the original author's data set has the same data for every 3 packets

并且原作者的数据集每3个数据包的数据都是相同的，如下图所示。

image

You trained the model with the original author's data and tested it yourself with good results. I do not understand how such a data set can predict blood sugar well, and I would like to ask how you processed the raw data to fit the original author's data set.

您使用原作者的数据训练模型，并自己采集数据进行测试获得了不错的效果。我不能理解这样的数据集为什么能预测好血糖，我想问一下您是如何处理原始数据以适配原作者的数据集。

Looking forward to your reply again!

再次期待您的答复！

**Comment by wmutian** [comment]
I am very sorry for not replying you in time because I am a little busy recently. Thank you very much for your waiting. Your questions are quite valuable, especially the data duplication you mentioned because I was just noticed that.

非常抱歉，最近事情比较多，未能及时回复您的留言，感谢您的耐心等待。您的问题很有价值，特别是您提到的数据重复，我也是最近才注意到。

Regarding the training of the model, the dataset used is not the original PPG data. This is because the data was normalized before being used for training, you can find it in the “glucose.ipynb” file provided by the original author, :

您提到训练模型时，数据集并非是原始的PPG数据，这是因为在原作者提供的glucose.ipynb文件中，数据用于训练前先做了标准化处理：

#数据规格化
    for i in range(seg):
        m = xdata[i].mean()
        s = xdata[i].std()
        xdata[i] -= m
        xdata[i] /= s
    ydata = []

The PPG data provided by the original author is sampled every 2 ms, while the data I used for testing is sampled every 20 ms. Below is the normalized PPG waveform of the data I collected, with the red line representing red light and the blue line representing infrared light. If the x-axis is stretched by a factor of ten, it might resemble the PPG waveform provided by the original author.

原作者提供的PPG数据每2 ms采集一次，我自己测试用的数据每20 ms采集一次，以下是我自己采集的数据标准化后的PPG波形，红色线条是红光，蓝色线条是红外光。如果将横坐标拉长十倍，可能和原作者提供的PPG波形是类似的。

Screenshot_20250303_235917_com tencent mobileqq_edit_90263881872163

Thank you for pointing out the issue of data duplication; I hadn't noticed it before. Recently, I've been reflecting on why such a dataset could be used to predict blood glucose levels. As far as I can tell, the multilayer perceptron (MLP) is like a multivariate nonlinear regression equation. After multiple rounds of training, it can obtain a set of parameters that fit the blood glucose levels relatively well, similar to an equation like y=ax^10+bx^9+…+c. It's likely that I just happened to train a set of parameters that fit well. Moreover, the blood glucose range for healthy individuals (3-9 mmol/L) is not very wide. The 4,042 data samples provided by the original author might be sufficient for training purposes.

关于数据集重复的问题，我之前确实没有注意到，谢谢您指出来这一点。最近我也在反思，这样的数据集为什么能预测出来血糖。我能想到的是多层感知机就像一个多元非线性回归方程，经过反复多轮的训练，能得到一个表现比较好的参数组合来拟合血糖，就像y=ax^10+bx^9+…+c一样，可能我只是碰巧训练出一个拟合较好的参数组合。而且正常人血糖范围3-9mmol/L也不算很大，原作者提供的4042组数据对于训练来说足够了。

I hope the above content is helpful to you. If you have any other questions,  feel free to contact me. I will try my best to help you.

希望以上对内容您有所帮助。如果您还有其他问题，欢迎继续留言，我会尽力帮您解答。
